File: onebot/src/main/java/cn/evole/onebot/sdk/util/json/MsgAdapter.java
package cn.evole.onebot.sdk.util.json;

import cn.evole.onebot.sdk.response.group.GetMsgResp;
import cn.evole.onebot.sdk.util.JsonHelper;
import com.google.gson.*;

import java.lang.reflect.Type;

public class MsgAdapter extends JsonHelper implements JsonDeserializer<GetMsgResp> {


    public GetMsgResp deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) {
        JsonObject obj = json.getAsJsonObject();

        int messageId = getMessageId(obj);
        int realId = ignorable(obj, "real_id", 0);
        GetMsgResp.Sender sender = gson.fromJson(obj.get("sender"), GetMsgResp.Sender.class);
        int time = obj.get("time").getAsInt();
        String message = forceString(obj, "message");
        String rawMessage = ignorable(obj, "raw_message", "");
        long peerId = obj.has("peer_id") ? obj.get("peer_id").getAsLong() : 0;
        long groupId = obj.has("group_id") ? obj.get("group_id").getAsLong() : 0;
        long targetId = obj.has("target_id") ? obj.get("target_id").getAsLong() : 0;

        return new GetMsgResp(messageId, realId, sender, time, message, rawMessage, peerId, groupId, targetId);
    }

    public static int getMessageId(JsonObject obj) {
        String msgId = obj.get("message_id").getAsString();
        try {
            return Integer.parseInt(msgId);
        } catch (NumberFormatException ignored) {
            throw new NumberFormatException(
                    "无法将消息ID `" + msgId + "` 的类型转换为 int32，" +
                    "请向你所使用的 Onebot 实现维护者报告该问题，" +
                    "不要将该问题反馈到 Overflow。"
            );
        }
    }
}

File: onebot/src/main/kotlin/sdk/util/json/MsgAdapter.kt
package cn.evolvefield.onebot.sdk.util.json

import cn.evolvefield.onebot.sdk.response.group.GetMsgResp
import cn.evolvefield.onebot.sdk.util.*
import com.google.gson.*
import java.lang.reflect.Type

class MsgAdapter : JsonDeserializerKt<GetMsgResp> {
    override fun deserializeFromJson(
        json: JsonElement,
        type: Type,
        ctx: JsonDeserializationContext
    ): GetMsgResp {
        val obj = json.asJsonObject
        return GetMsgResp().apply {
            messageId = obj.messageId()
            realId = obj.ignorable("real_id", 0)
            sender = obj.fromJson("sender")!!
            time = obj.int("time")
            isJsonMessage = obj.has("message") && obj["message"].isJsonArray
            message = obj.forceString("message")
            rawMessage = obj.ignorable("raw_message", "")
            peerId = obj.ignorable("peer_id", 0L)
            groupId = obj.ignorable("group_id", 0L)
            targetId = obj.ignorable("target_id", 0L)
        }
    }

    companion object {
        fun JsonObject.messageId(): Int {
            val msgId: String = this["message_id"]?.asString ?: throw IllegalStateException("""
                收到消息ID为空的消息事件。type=${this["message_type"]}, sub_type=${this["sub_type"]}
                请向你所使用的 Onebot 实现维护者报告该问题，
                不要将该问题反馈到 Overflow。
            """.trimIndent()
            )
            try {
                return msgId.toInt()
            } catch (ignored: NumberFormatException) {
                throw NumberFormatException(
                    """
                        无法将消息ID $msgId 的类型转换为 int32，
                        请向你所使用的 Onebot 实现维护者报告该问题，
                        不要将该问题反馈到 Overflow。
                    """.trimIndent()
                )
            }
        }
    }
}